nimare/meta/cbma/mkda.py

```diff
@@ -518,168 +488,92 @@ def _compute_summarystat(self, data):
-            assert "empirical_null" in self.null_distributions_.keys()
-            p_values = null_to_p(
-                stat_values, self.null_distributions_["empirical_null"], tail="upper"
-            )
+        if isinstance(ma_maps, list):
+            ma_values = self.masker.transform(ma_maps)
+        elif isinstance(ma_maps, np.ndarray):
+            ma_values = ma_maps.copy()
         else:
-            raise ValueError("Argument 'null_method' must be one of: 'analytic', 'empirical'.")
-
-        z_values = p_to_z(p_values, tail="one")
-        return p_values, z_values
+            raise ValueError('Unsupported data type "{}"'.format(type(ma_maps)))
 
-    def _run_fwe_permutation(self, params):
-        """
-        Run a single Monte Carlo permutation of a dataset. Does vFWE, but not cFWE.
-        """
-        iter_ijk, iter_df = params
-        iter_ijk = np.squeeze(iter_ijk)
-        iter_df[["i", "j", "k"]] = iter_ijk
-        iter_ma_maps = self.kernel_transformer.transform(
-            iter_df, masker=self.masker, return_type="array"
+        # assumes that groupby results in same order as MA maps
+        n_foci_per_study = self.inputs_["coordinates"].groupby("id").size().values
+
+        # Determine bins for null distribution histogram
+        # The maximum possible MA value for each study is the weighting factor (generally 1)
+        # times the number of foci in the study.
+        # To get the weighting factor, we find the minimum value in each MA map, ignoring zeros.
+        n_studies = ma_values.shape[0]
+        min_ma_values = np.zeros(n_studies)
+        for i_study in range(n_studies):
+            temp_ma_values = ma_values[i_study, :]
+            min_ma_values[i_study] = np.min(temp_ma_values[temp_ma_values != 0])
+
+        step_size = np.mean(min_ma_values)  # typically 1
+        inv_step_size = int(np.ceil(1 / step_size))  # only useful when weight != 1
+
+        max_ma_values = min_ma_values * n_foci_per_study
+        max_poss_value = self._compute_summarystat(max_ma_values)
+        # Weighting is not supported yet, so I'm going to build my bins around the min MA value.
+        # Remember that numpy histogram bins are bin edges, not centers.
+        # Assuming values of 0, 1, 2, etc., bins are -0.5-0.5, 0.5-1.0, etc.
+        hist_bins = np.arange(
+            -step_size / 2, max_poss_value + (step_size * 1.5) + 0.001, step_size
         )
-        iter_of_map = np.sum(iter_ma_maps, axis=0)
-        iter_max_value = np.max(iter_of_map)
-        return iter_max_value
 
-    def correct_fwe_montecarlo(self, result, n_iters=10000, n_cores=-1):
-        """
-        Perform FWE correction using the max-value permutation method.
-        Only call this method from within a Corrector.
+        def just_histogram(*args, **kwargs):
+            """Collect the first output (weights) from numpy histogram."""
+            return np.histogram(*args, **kwargs)[0].astype(float)
```

Comment:
We could use this with ALE too.

Reply:
I know that `apply_along_axis` is just a for loop, but I figured this made for cleaner code, even if it's not any faster.

Reply:
Implemented in ALE in #411.